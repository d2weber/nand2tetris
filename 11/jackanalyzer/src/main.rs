mod compilation_engine;
mod token;

use std::{env, path::PathBuf};

fn main() {
    let mut args = env::args();
    let path = if args.len() == 1 {
        // Default to current directory
        env::current_dir().unwrap()
    } else if args.len() == 2 {
        let filename = args.next_back().unwrap();
        PathBuf::from(&filename)
    } else {
        panic!("Zero parameters or one parameter expected.");
    };
    compilation_engine::compile_path(path.as_path()).unwrap();
}

#[cfg(test)]
mod test {
    use std::{fs, io::BufWriter, path::Path};

    use super::*;

    #[test]
    fn seven() {
        snapshot_directory("../Seven");
    }

    #[test]
    fn convert_to_bin() {
        snapshot_directory("../ConvertToBin");
    }

    #[test]
    fn square() {
        snapshot_directory("../Square");
    }

    #[test]
    fn average() {
        snapshot_directory("../Average");
    }

    #[test]
    fn pong() {
        snapshot_directory("../Pong");
    }

    #[test]
    fn complex_array() {
        snapshot_directory("../ComplexArrays");
    }

    #[test]
    fn snake() {
        snapshot_directory("../../9/Snake");
    }

    fn snapshot_directory(s: &str) {
        let path = Path::new(s);
        let cargo_root = Path::new(env!("CARGO_MANIFEST_DIR"));
        let path = cargo_root.join(path);

        for file in files_with_extension(&path, "jack") {
            let snapshot_name = file.strip_prefix(cargo_root).unwrap().display().to_string();
            snapshot_compiled(&snapshot_name, &file);
        }
    }

    fn snapshot_compiled(snapshot_name: &str, file_name: &Path) {
        let mut out = BufWriter::new(Vec::new());
        compilation_engine::compile_file(file_name, &mut out);
        let vm_code = String::from_utf8(out.into_inner().unwrap()).unwrap();
        insta::assert_snapshot!(snapshot_name, vm_code);
    }

    fn files_with_extension<'a>(
        path: &'a Path,
        extension: &'a str,
    ) -> impl Iterator<Item = PathBuf> + 'a {
        fs::read_dir(path).unwrap().filter_map(move |dir_entry| {
            let file = dir_entry.unwrap().path();
            file.extension()
                .is_some_and(|e| e == extension)
                .then_some(file)
        })
    }
}
