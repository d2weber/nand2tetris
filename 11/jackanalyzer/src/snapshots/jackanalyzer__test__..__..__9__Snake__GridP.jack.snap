---
source: src/main.rs
expression: vm_code
---
function GridP.new 0
push constant 2
call Memory.alloc 1
pop pointer 0
push argument 0
pop this 0
push argument 1
pop this 1
push pointer 0
call GridP.draw 1
pop temp 0
push pointer 0
return
function GridP.maybeNew 0
push argument 0
push constant 0
lt
push argument 0
call GridP.max_x 0
gt
or
push argument 1
push constant 0
lt
or
push argument 1
call GridP.max_y 0
gt
or
not
if-goto IF_FALSE0
push constant 0
return
goto IF_TRUE1
label IF_FALSE0
label IF_TRUE1
push argument 0
push argument 1
call GridP.new 2
return
function GridP.x 0
push argument 0
pop pointer 0
push this 0
return
function GridP.y 0
push argument 0
pop pointer 0
push this 1
return
function GridP.eq 0
push argument 0
pop pointer 0
push this 0
push argument 1
call GridP.x 1
eq
push this 1
push argument 1
call GridP.y 1
eq
and
return
function GridP.left 0
push argument 0
pop pointer 0
push this 0
push constant 1
sub
push this 1
call GridP.maybeNew 2
return
function GridP.right 0
push argument 0
pop pointer 0
push this 0
push constant 1
add
push this 1
call GridP.maybeNew 2
return
function GridP.above 0
push argument 0
pop pointer 0
push this 0
push this 1
push constant 1
sub
call GridP.maybeNew 2
return
function GridP.below 0
push argument 0
pop pointer 0
push this 0
push this 1
push constant 1
add
call GridP.maybeNew 2
return
function GridP.draw 0
push argument 0
pop pointer 0
push constant 1
neg
call Screen.setColor 1
pop temp 0
push this 0
push constant 16
call Math.multiply 2
push this 1
push constant 16
call Math.multiply 2
push this 0
push constant 1
add
push constant 16
call Math.multiply 2
push constant 1
sub
push this 1
push constant 1
add
push constant 16
call Math.multiply 2
push constant 1
sub
call Screen.drawRectangle 4
pop temp 0
push constant 0
return
function GridP.undraw 0
push argument 0
pop pointer 0
push constant 0
call Screen.setColor 1
pop temp 0
push this 0
push constant 16
call Math.multiply 2
push this 1
push constant 16
call Math.multiply 2
push this 0
push constant 1
add
push constant 16
call Math.multiply 2
push constant 1
sub
push this 1
push constant 1
add
push constant 16
call Math.multiply 2
push constant 1
sub
call Screen.drawRectangle 4
pop temp 0
push constant 0
return
function GridP.disposeKeepDrawn 0
push argument 0
pop pointer 0
push pointer 0
call Memory.deAlloc 1
pop temp 0
push constant 0
return
function GridP.dispose 0
push argument 0
pop pointer 0
push pointer 0
call GridP.undraw 1
pop temp 0
push pointer 0
call Memory.deAlloc 1
pop temp 0
push constant 0
return
function GridP.max_x 0
push constant 31
return
function GridP.max_y 0
push constant 15
return
